### PR TITLE
Port Reachable API

### PR DESCRIPTION
**Pull request by SanjuPal01**
### :nut_and_bolt: Description: What code changed, and why?
Created /api/v1/checks/port-reachable API, which will check that given ip is reachable on the given port or not. 

### :chains: Related Resources
https://chefio.atlassian.net/browse/CHEF-2294

### :+1: Definition of Done
Implemented this API as mentioned in the ticket.

### :athletic_shoe: How to Build and Test the Change
This is the url: http://localhost:7799/api/v1/checks/port-reachable and request method is POST

### :white_check_mark: Checklist

**All PRs** must tick these:

- [x] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [x] All commits signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

With occasional exceptions, all PRs **from Progress employees** must tick these:

- [x] Is the code clear? *(complicated code or lots of comments--subdivide and use well-named methods, meaningful variable names, etc.)*
- [x] Consistency checked? *(user notifications, user prompts, visual patterns, code patterns, variable names)*
- [x] Repeated code blocks eliminated? *(adapt and reuse existing components, blocks, functions, etc.)*
- [x] Spelling, grammar, typos checked? *(at a minimum use `make spell` in any component directory)*
- [x] Code well-formatted? *(indents, line breaks, etc. improve rather than hinder readability)*

All PRs **from Progress employees** should tick these if appropriate:

- [x] Tests added/updated? (all new code needs new tests)
- [ ] Docs added/updated? (all customer-facing changes)

*Please add a note next to any checkbox above if you are NOT ticking it.*

### :camera: Screenshots, if applicable
Screenshot 2023-05-22 at 6 23 28 PM
Screenshot 2023-05-22 at 6 20 37 PM
Screenshot 2023-05-22 at 6 20 57 PM
Screenshot 2023-05-22 at 6 25 51 PM
Screenshot 2023-05-22 at 6 26 06 PM
Screenshot 2023-05-22 at 6 27 30 PM
Screenshot 2023-05-22 at 6 32 12 PM
Screenshot 2023-05-22 at 6 27 53 PM
Screenshot 2023-05-22 at 6 28 20 PM
Screenshot 2023-05-22 at 6 28 49 PM

### CODE REVIEW

**Review by Dmaddu — CHANGES_REQUESTED** [review]
please fix the pipelines

File: components/automate-cli/pkg/verifyserver/services/portreachableservice/portreachable.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+	}
+	pr.log.Debug("Server Response(if any): ", string(buffer[:n]))
+	return nil
+}
+
+func (pr *PortReachableService) httpsClient(ip, cert string, port int) error {
+	caCertPool := x509.NewCertPool()
+	ok := caCertPool.AppendCertsFromPEM([]byte(cert))
+	if !ok {
+		return errors.New("certificate error")
```
Comment (YashviJain01) [inline components/automate-cli/pkg/verifyserver/services/portreachableservice/portreachable.go:89]: What is the certificate error which comes here, can we have some valid error?

File: components/automate-cli/pkg/verifyserver/services/portreachableservice/portreachable.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+	n, _, err := conn.ReadFromUDP(buffer)
+	if err != nil {
+		return fmt.Errorf("read error: %w", err)
+	}
+	pr.log.Debug("Server Response(if any): ", string(buffer[:n]))
+	return nil
+}
+
+func (pr *PortReachableService) httpsClient(ip, cert string, port int) error {
+	caCertPool := x509.NewCertPool()
```
Comment (YashviJain01) [inline components/automate-cli/pkg/verifyserver/services/portreachableservice/portreachable.go:86]: Add debugger log here as well, trying to connect to https server?

File: components/automate-cli/pkg/verifyserver/services/portreachableservice/portreachable.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+	if err != nil {
+		return fmt.Errorf("dial failed: %w", err)
+	}
+	pr.log.Debug("Got the Connection Object")
+	conn.Close()
+	return nil
+}
+
+func (pr *PortReachableService) udpClient(protocolType, host, port string) error {
+	pr.log.Debug("URL: ", net.JoinHostPort(host, port))
```
Comment (YashviJain01) [inline components/automate-cli/pkg/verifyserver/services/portreachableservice/portreachable.go:47]: Debugger logs here as well.

**Review by kalroy — CHANGES_REQUESTED** [review]

File: components/automate-cli/pkg/verifyserver/server/api/v1/portreachable.go [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+		c.Next(&fiber.Error{Code: http.StatusBadRequest, Message: "Invalid Body Request"})
+		return
+	}
+	if reqBody.DestinationNodeIp == "" ||
+		reqBody.DestinationNodePort == 0 ||
+		reqBody.DestinationNodeServiceProtocol == "" {
+		c.Next(&fiber.Error{Code: http.StatusBadRequest, Message: "DestinationNodeIp, DestinationNodePort, or DestinationNodeServiceProtocol cannot be empty"})
+		return
+	}
+	if reqBody.DestinationNodeServiceProtocol == constants.HTTPS && reqBody.RootCA == "" {
```
[the remark this answers is not in the document]
Comment (SanjuPal01) [reply components/automate-cli/pkg/verifyserver/server/api/v1/portreachable.go:25]: We are getting only DestinationIP in request body. Hence we can't do this check. Firewall API need to perform this check.

File: components/automate-cli/pkg/verifyserver/server/api/v1/portreachable.go [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+		c.Next(&fiber.Error{Code: http.StatusBadRequest, Message: "please give rootCA value"})
```
Comment (vivekshankar1) [inline components/automate-cli/pkg/verifyserver/server/api/v1/portreachable.go:26]:
```suggestion
		c.Next(&fiber.Error{Code: http.StatusBadRequest, Message: "RootCA value is mandatory for protocol HTTPS"})
```

File: components/automate-cli/pkg/verifyserver/services/portreachableservice/portreachable.go [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+func NewPortReachableService(log logger.Logger, timeout time.Duration) IPortReachableService {
+	return &PortReachableService{
+		log:     log,
+		timeout: timeout,
+	}
+}
+
+func (pr *PortReachableService) tcpClient(protocolType, host, port string) error {
+	pr.log.Debug("Trying to connect TCP server")
```
Comment (vivekshankar1) [inline components/automate-cli/pkg/verifyserver/services/portreachableservice/portreachable.go:34]: please add some information with debug logs as well like protocol, host, port etc

File: components/automate-cli/pkg/verifyserver/server/api/v1/portreachable.go [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+func (h *Handler) PortReachable(c *fiber.Ctx) {
+	reqBody := models.PortReachableRequest{}
+	if err := c.BodyParser(&reqBody); err != nil {
+		h.Logger.Error(err.Error())
+		c.Next(&fiber.Error{Code: http.StatusBadRequest, Message: "Invalid Body Request"})
+		return
+	}
+	if reqBody.DestinationNodeIp == "" ||
+		reqBody.DestinationNodePort == 0 ||
```
[the remark this answers is not in the document]
Comment (SanjuPal01) [reply components/automate-cli/pkg/verifyserver/server/api/v1/portreachable.go:20]: I added one extra check for port range. now this check only handle if user didn't give any port number then default int value 0 will be passed and this block will be executed. 
but if user gave some invalid port number like negative number then i am handling that seperately in the next if block.

**Review by Sahiba3108 — APPROVED** [review]

**Review by kalroy — APPROVED** [review]

**Review by Dmaddu — APPROVED** [review]

**Pull request merged**